pkg/summary/summary.go

```diff
@@ -49,119 +58,144 @@ type Summary struct {
+		excludedContainers := sets.NewString().Union(s.excludedContainers)
+		if val, exists := deployment.GetAnnotations()[deploymentExcludeContainersAnnotation]; exists {
+			excludedContainers.Insert(strings.Split(val, ",")...)
 		}
 
 	CONTAINER_REC_LOOP:
 		for _, containerRecommendation := range vpa.Status.Recommendation.ContainerRecommendations {
-			for _, exclusion := range containerExclusions {
-				if exclusion == containerRecommendation.ContainerName {
-					klog.V(2).Infof("Excluding container %v", containerRecommendation.ContainerName)
-					continue CONTAINER_REC_LOOP
-				}
+			if excludedContainers.Has(containerRecommendation.ContainerName) {
+				klog.V(2).Infof("Excluding container Deployment/%s/%s", dSummary.DeploymentName, containerRecommendation.ContainerName)
+				continue CONTAINER_REC_LOOP
 			}
 
-			var container = containerSummary{
-				ContainerName:  containerRecommendation.ContainerName,
-				UpperBound:     utils.FormatResourceList(containerRecommendation.UpperBound),
-				LowerBound:     utils.FormatResourceList(containerRecommendation.LowerBound),
-				Target:         utils.FormatResourceList(containerRecommendation.Target),
-				UncappedTarget: utils.FormatResourceList(containerRecommendation.UncappedTarget),
-			}
+			var cSummary containerSummary
 			for _, c := range deployment.Spec.Template.Spec.Containers {
 				if c.Name == containerRecommendation.ContainerName {
-					klog.V(6).Infof("Resources for %s: %v", c.Name, c.Resources)
-					container.Limits = utils.FormatResourceList(c.Resources.Limits)
-					container.Requests = utils.FormatResourceList(c.Resources.Requests)
-					break
+					cSummary = containerSummary{
+						ContainerName:  containerRecommendation.ContainerName,
+						UpperBound:     utils.FormatResourceList(containerRecommendation.UpperBound),
+						LowerBound:     utils.FormatResourceList(containerRecommendation.LowerBound),
+						Target:         utils.FormatResourceList(containerRecommendation.Target),
+						UncappedTarget: utils.FormatResourceList(containerRecommendation.UncappedTarget),
+					}
+					cSummary.Limits = utils.FormatResourceList(c.Resources.Limits)
+					cSummary.Requests = utils.FormatResourceList(c.Resources.Requests)
+					klog.V(6).Infof("Resources for Deployment/%s/%s: Requests: %v Limits: %v", dSummary.DeploymentName, c.Name, cSummary.Requests, cSummary.Limits)
 				}
 			}
 
-			deploy.Containers = append(deploy.Containers, container)
+			dSummary.Containers = append(dSummary.Containers, cSummary)
 		}
-		summary.Deployments = append(summary.Deployments, deploy)
+		summary.Deployments = append(summary.Deployments, dSummary)
 	}
 
-	summary.Namespaces = utils.UniqueString(summary.Namespaces)
+	// get the unique list of namespaces we've seen for this summary
+	summary.Namespaces = summaryNamespaces.List()
+
 	return summary, nil
 }
+
+// Update the list of VPAs that the summarizer uses
+func (s *Summarizer) UpdateVPAs() error {
```

Comment:
Does this need to be a public method? It looks like the only caller is in this package.

Reply:
No, I catch this in an upcoming commit and change this to be private as well as revisit the caching behaviour of the Summarizer object to make a little more sense from a public API perspective.